### DISCUSSION TITLE
Precedence of repo config with `extends`

### DISCUSSION THREAD

**Discussion by adam-moss** [renovatebot/renovate#19384] [Request Help]
Can I clarify my understanding of how Renovate should act when

1) there is a global `config.yaml` set via `RENOVATE_CONFIG_FILE`
2) it contains `extends:`
3) it also contains `packageRules:`, a number of which set `automerge: true`
4) the repository has a `renovate.json`
5) it contains `extends:`, which is setting `automerge: false` for the global `packageRules`

Looking at the merged config I can see

```json
"config": {
  ~redact~
  "extends": [
    "workarounds:all",
    "npm:unpublishSafe",
    "default:ignoreUnstable",
    "default:enablePreCommit",
    "gitlab>~redact~/~redact~/renovate//shared-configs/whitesource-merge-confidence",
    "gitlab>~redact~/~redact~/renovate//shared-configs/~redact~",
    "gitlab>~redact~/~redact~/renovate//shared-configs/disable-default-automerge"
  ],
  "packageRules": [
    ~redact~
  ]
}
``` 

That last line in the `extends` array coming from the repository `renovate.json` and everything else from the global `config.yaml`

Would I be right in thinking the reason the things in `disable-default-automerge` are still being automerged is because the `packageRules` are processed afterwards, effectively nullifying the `extends` content?

If that is the case should those `packageRules` be moved into an shared-config themselves, should the `extends`  directive be the last entry in the global `config.yaml`, or something else?

Thanks!

**Comment by rarkins** [comment]
There is some logic to how this works, although it's not perfect. See #4726 for more details and thoughts on the topic.

I'll try to describe it enough here:

If you want presets to _resolve_ immediately in bot config, use `globalExtends`. If you want them to be passed through and resolved as part of repo config then use `extends`.

Package rules take precedence over non-package rules.

Repo config takes precedence over global config, but still packageRules get evaluated last.

> * there is a global `config.yaml` set via `RENOVATE_CONFIG_FILE`

Sure, this is global config. As a general rule anything here can get overridden by repo config later.

> * it contains `extends:`

These will essentially be added to repo config.

> * it also contains `packageRules:`, a number of which set `automerge: true`

Also part of global config I assume? If so then these rules can override any root automerge setting.

> * the repository has a `renovate.json`
> * it contains `extends:`, which is setting `automerge: false` for the global `packageRules`

Ultimately, the packageRules wins. The precedence is kind of like this (last wins):

global.automerge
repo.automerge
global.packageRules.automerge
repo.packageRules.automerge

**Reply by adam-moss** [reply]
Thanks for that explanation.  I thought that would be case but isn't what I'm seeing.  What I'm seeing is

global.automerge
repo.automerge
repo.packageRules.automerge (from extends)
global.packageRules.automerge

I think the answer here is for me to move those global packagerules into a separate shared config.

**Reply by rarkins** [reply]
Sorry, I think you're right about that. It's another nuance of the merging of config. Direct packageRules take priority over packageRules from extends (rationale: you define a packageRule to override some nested packageRule inside an extends). So then you're seeing that global packageRules are being merged with repo packageRules, and then extends packageRules are being put *before* (lower priority).

This is something I would consider changing though, but overall my recommendation would be to use shared configs and avoid "repo config" in global config wherever possible.

**Reply by adam-moss** [reply]
Yeah, added as as separate shared-config and that seems to have worked.

Thanks for the detailed explanation :+1:
